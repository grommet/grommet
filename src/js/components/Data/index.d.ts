import * as React from 'react';
import { BoxProps } from '../Box/index';

export interface View {
  name?: string;
  page?: number; // when paging
  properties?:
    | string[]
    | {
        [property: string]:
          | undefined
          | boolean
          | (string | number | boolean)[]
          | { max: number; min: number };
      };
  search?: string;
  sort?: {
    property: string;
    direction: 'asc' | 'desc';
  };
  step?: number; // page size
  columns?: string[];
}

export interface DataProps {
  id?: string;

  data: object[];
  total?: number;
  // when paging
  filteredTotal?: number;

  defaultView?: View;
  view?: string | View;
  onView?: (view: View) => void;

  // whether to render a Toolbar
  toolbar?: boolean | 'search' | 'filters' | 'view';

  properties?:
    | string[]
    | {
        [key: string]: {
          badge?: boolean;
          filter?: boolean;
          // for DataTable column header, DataFilter label, DataTableColumns label
          label?: string | React.ReactNode;
          // DataFilter options
          options?: (
            | string
            | number
            | {
                label: string;
                value: string | number | boolean;
              }
          )[];
          // DataFilter range
          range?: {
            max?: number;
            min?: number;
            step?: number;
          };
          // for internal filtering only, should searching evaluate this property
          search?: boolean;
          sort?: boolean;
        };
      };

  messages?: {
    dataFilters?: {
      clear?: string;
      heading?: string;
      open?: string;
      openSet?: {
        singular?: string;
        plural?: string;
      };
    };
    dataForm?: {
      submit?: string;
    };
    dataSearch?: {
      label?: string;
      open?: string;
    };
    dataSort?: {
      ascending?: string;
      by?: string;
      descending?: string;
      direction?: string;
      open?: string;
    };
    dataSummary?: {
      filtered?: string;
      filteredSingle?: string;
      items?: string;
      itemsSingle?: string;
      selected?: string;
      total?: string;
      totalSingle?: string;
    };
    dataTableColumns?: {
      open?: string;
      order?: string;
      select?: string;
      tip?: string;
    };
    dataTableGroupBy?: {
      clear?: string;
      label?: string;
    };
    dataView?: {
      label?: string;
    };
  };

  views?: View[];

  // Future selection
  // select: string[];
  // onSelect: (select: string[]) => void;

  // Future manage named views
  // onCreateView?: (view: View) =>  void;
  // onDeleteView?: (string) => void;
}

type divProps = Omit<
  React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement>,
  'onClick'
>;

export interface DataExtendedProps extends BoxProps, DataProps, divProps {}

declare const Data: React.FC<DataExtendedProps>;

export { Data };
