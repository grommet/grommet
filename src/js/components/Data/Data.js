import React, { useContext, useEffect, useMemo, useState } from 'react';
import { AnnounceContext } from '../../contexts';
import { Box } from '../Box';
import { DataFilters } from '../DataFilters';
import { DataSearch } from '../DataSearch';
import { DataSummary } from '../DataSummary';
import { DataView } from '../DataView';
import { Toolbar } from '../Toolbar';
import { DataContext } from '../../contexts/DataContext';
import { DataPropTypes } from './propTypes';
import { MessageContext } from '../../contexts/MessageContext';
import { filter } from './filter';

const defaultDefaultView = { search: '' };

const normalizeView = (viewProp, views) =>
  (typeof viewProp === 'string' && views?.find((v) => v.name === viewProp)) ||
  (typeof viewProp === 'object' && viewProp);

export const Data = ({
  children,
  data: dataProp = [],
  defaultView = defaultDefaultView,
  filteredTotal,
  id = 'data',
  messages,
  onView,
  properties,
  toolbar,
  total,
  view: viewProp,
  views,
  ...rest
}) => {
  const [view, setView] = useState(
    normalizeView(viewProp || defaultView, views),
  );
  useEffect(
    () => setView(normalizeView(viewProp || defaultView, views)),
    [defaultView, viewProp, views],
  );
  const [toolbarKeys, setToolbarKeys] = useState([]);

  const result = useMemo(() => {
    if (onView)
      // caller is filtering
      return {
        data: dataProp,
        total,
        filteredTotal: filteredTotal ?? dataProp?.length ?? 0,
      };
    return filter(dataProp, view, properties);
  }, [dataProp, filteredTotal, onView, properties, total, view]);

  // used by DataFilters to determine if badge should appear on Filter button
  const [filtersCleared, setFiltersCleared] = useState(true);

  const [selected, setSelected] = useState([]);

  const announce = useContext(AnnounceContext);
  const { format } = useContext(MessageContext);
  // Announce to screen readers when search or filters are
  // applied and affect the underlying result set
  useEffect(() => {
    let messageId;
    if (result.total !== result.filteredTotal) {
      if (result.filteredTotal === 1) messageId = 'dataSummary.filteredSingle';
      else messageId = 'dataSummary.filtered';
    } else if (result.total === 1) messageId = 'dataSummary.totalSingle';
    else messageId = 'dataSummary.total';

    // helps account for cases like 0 results of 1 item
    const items = format({
      id: result.total === 1 ? 'dataSummary.itemsSingle' : 'dataSummary.items',
      messages: messages?.dataSummary,
    });

    announce(
      `${format({
        id: messageId,
        messages: messages?.dataSummary,
        values: {
          filteredTotal: result.filteredTotal,
          total: result.total,
          items,
        },
      })}${
        selected > 0
          ? `, ${format({
              id: 'dataSummary.selected',
              messages: messages?.dataSummary,
              values: {
                selected,
              },
            })}`
          : ''
      }`,
    );
  }, [
    announce,
    format,
    messages?.dataSummary,
    result.filteredTotal,
    result.total,
    selected,
  ]);

  // what we use for DataContext value
  const contextValue = useMemo(() => {
    const value = {
      id,
      messages,
      properties,
      filtersCleared,
      setFiltersCleared,
      selected,
      setSelected,
      view,
      views,
      ...result,
    };

    value.clearFilters = () => {
      const nextView = { ...view };
      delete nextView.properties;
      delete nextView.page;
      // by clearing the properties from a view, it is no
      // longer reflecting the view
      delete nextView.name;
      delete nextView.view;
      setFiltersCleared(true);
      setView(nextView);
      if (onView) onView(nextView);
    };

    value.onView = (nextView) => {
      setView(nextView);
      if (onView) onView(nextView);
    };

    value.addToolbarKey = (key) => {
      setToolbarKeys((prevKeys) => {
        if (prevKeys.includes(key)) return prevKeys;
        return [...prevKeys, key];
      });
    };
    value.toolbarKeys = toolbarKeys;

    return value;
  }, [
    id,
    messages,
    filtersCleared,
    onView,
    properties,
    result,
    selected,
    toolbarKeys,
    view,
    views,
  ]);

  let toolbarContent;
  if (toolbar) {
    toolbarContent = [
      // TO DO theme object
      <Toolbar key="toolbar" gap="medium">
        <Toolbar>
          {(toolbar === true || toolbar === 'search') && <DataSearch />}
          {(toolbar === true || toolbar === 'filters') && <DataFilters layer />}
        </Toolbar>
        {(toolbar === true || toolbar === 'view') && <DataView />}
      </Toolbar>,
      <DataSummary key="summary" />,
    ];
  }

  return (
    <DataContext.Provider value={contextValue}>
      <Box id={id} flex={false} {...rest}>
        {toolbarContent}
        {children}
      </Box>
    </DataContext.Provider>
  );
};

Data.propTypes = DataPropTypes;
