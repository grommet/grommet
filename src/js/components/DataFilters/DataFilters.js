import React, {
  Children,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { Filter } from 'grommet-icons/icons/Filter';
import { Close } from 'grommet-icons/icons/Close';
import { Box } from '../Box';
import { Button } from '../Button';
import { DataClearFilters } from '../DataClearFilters';
import { DataFilter } from '../DataFilter';
import { DataForm } from '../Data/DataForm';
import { DropButton } from '../DropButton';
import { Header } from '../Header';
import { Heading } from '../Heading';
import { Layer } from '../Layer';
import { DataContext } from '../../contexts/DataContext';
import { MessageContext } from '../../contexts/MessageContext';
import { DataFiltersPropTypes } from './propTypes';
import { useThemeValue } from '../../utils/useThemeValue';

const dropProps = {
  align: { top: 'bottom', right: 'right' },
};

const layerProps = {
  full: 'vertical',
  position: 'right',
};

const defaultTouched = {};
export const DataFilters = ({
  drop,
  children,
  clearFilters = true,
  heading,
  layer,
  updateOn,
  ...rest
}) => {
  const {
    id: dataId,
    messages,
    properties,
    unfilteredData,
    filtersCleared,
    setFiltersCleared,
    view,
  } = useContext(DataContext);
  const { format } = useContext(MessageContext);
  const { theme } = useThemeValue();
  const [showContent, setShowContent] = useState();
  // touched is a map of property to its value based on if user interacts
  // with a filter or a view applies of set of filters
  const [touched, setTouched] = useState(defaultTouched);

  // if filters have been applied by this DataFilters, update
  // the DataContext that filters are not in a "cleared" state
  useEffect(() => {
    setFiltersCleared(!Object.keys(touched).length);
  }, [touched, setFiltersCleared]);

  // if filters have been cleared via clearFilters in DataContext,
  // reset touched to default state so badge is removed
  useEffect(() => {
    if (filtersCleared) {
      setTouched(defaultTouched);
    }
  }, [filtersCleared]);
  const controlled = useMemo(() => drop || layer, [drop, layer]);

  const configured = Children.count(children) === 0;
  useEffect(() => {
    // when view changes via DataView or user interacting with filters,
    // adjust badge to reflect that
    if (controlled && view.properties) {
      const nextTouched = { ...view.properties };
      Object.keys(nextTouched).forEach((k) => {
        if (
          properties?.[k]?.badge === false ||
          (configured && properties && !properties?.[k])
        ) {
          delete nextTouched[k];
        }
      });
      setTouched(nextTouched);
    }
  }, [configured, controlled, properties, view]);

  // generate the badge value based on touched fields that have a value.
  // only show the badge based off of what's included in this DataFilters
  // since multiple DataFilters may exist
  const badge = useMemo(
    () =>
      (controlled && Object.keys(touched).filter((k) => touched[k]).length) ||
      undefined,
    [controlled, touched],
  );

  const clearControl = badge && clearFilters && (
    <Box flex={false} margin={theme.dataFilters.clearControl?.margin}>
      <DataClearFilters />
    </Box>
  );

  let content = children;
  if (Children.count(children) === 0) {
    let filtersFor;
    if (!properties && unfilteredData && unfilteredData.length)
      // build from a piece of data, ignore object values
      filtersFor = Object.keys(unfilteredData[0]).filter(
        (k) => typeof unfilteredData[0][k] !== 'object',
      );
    else if (Array.isArray(properties)) filtersFor = properties;
    else if (typeof properties === 'object') {
      filtersFor = Object.keys(properties).filter(
        (property) => !(properties[property]?.filter === false),
      );
    } else filtersFor = [];
    content = filtersFor.map((property) => (
      <DataFilter key={property} property={property} />
    ));
  }

  content = (
    <DataForm
      // TO DO theme object
      pad={controlled ? 'medium' : undefined}
      onDone={() => setShowContent(false)}
      updateOn={updateOn}
      {...(!controlled ? rest : { fill: 'vertical' })}
    >
      {layer && (
        <Header>
          <Heading margin="none" level={2}>
            {heading ||
              format({
                id: 'dataFilters.heading',
                messages: messages?.dataFilters,
              })}
          </Heading>
          {!controlled && clearControl}
          <Button icon={<Close />} onClick={() => setShowContent(undefined)} />
        </Header>
      )}
      {content}
    </DataForm>
  );

  if (!controlled) return content;

  const tip = format({
    id: badge
      ? `dataFilters.openSet.${badge === 1 ? 'singular' : 'plural'}`
      : 'dataFilters.open',
    messages: messages?.dataFilters,
    values: { number: badge },
  });

  let control;
  if (drop) {
    control = (
      <DropButton
        id={`${dataId}--filters-control`}
        tip={tip}
        aria-label={tip}
        kind={theme.data.button?.kind}
        icon={<Filter />}
        dropProps={dropProps}
        dropContent={content}
        badge={badge}
        open={showContent}
        onOpen={() => setShowContent(undefined)}
        onClose={() => setShowContent(undefined)}
      />
    );
  } else if (layer) {
    control = (
      <Button
        id={`${dataId}--filters-control`}
        tip={tip}
        aria-label={tip}
        kind={theme.data.button?.kind}
        icon={<Filter />}
        badge={badge}
        onClick={() => setShowContent(true)}
      />
    );
  }

  return (
    <Box flex={false} direction="row" {...rest}>
      {control}
      {clearControl}
      {layer && showContent && (
        <Layer
          id={`${dataId}--filters-layer`}
          {...(typeof layer === 'object' ? layer : layerProps)}
          onClickOutside={() => setShowContent(undefined)}
          onEsc={() => setShowContent(undefined)}
        >
          <Box width={theme.dataFilters.content?.width}>{content}</Box>
        </Layer>
      )}
    </Box>
  );
};

DataFilters.propTypes = DataFiltersPropTypes;
