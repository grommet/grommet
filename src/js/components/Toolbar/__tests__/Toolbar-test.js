import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import { hpe } from 'grommet-theme-hpe';

import { Data } from '../../Data';
import { DataFilters } from '../../DataFilters';
import { DataSearch } from '../../DataSearch';
import { Grommet } from '../../Grommet';
import { Toolbar } from '..';

const data = [{ name: 'a' }, { name: 'b' }];

describe('DataFilters', () => {
  test('renders', () => {
    const { container } = render(
      <Grommet theme={hpe}>
        <Data data={data}>
          <Toolbar>
            <DataSearch />
            <DataFilters />
          </Toolbar>
        </Data>
      </Grommet>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
  test('renders outside grommet wrapper', () => {
    const { container } = render(
      <Data data={data}>
        <Toolbar>
          <DataSearch />
          <DataFilters />
        </Toolbar>
      </Data>,
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
