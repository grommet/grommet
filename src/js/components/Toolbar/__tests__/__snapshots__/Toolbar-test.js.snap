// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataFilters renders 1`] = `
.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c3 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c9 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 100%;
}

.c10 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 1 1;
  overflow: auto;
}

.c11 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-bottom: 12px;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c13 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  padding: 12px;
}

.c15 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-right: 12px;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  justify-content: center;
}

.c18 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  align-items: center;
  border: solid 2px rgba(0, 0, 0, 0.15);
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 24px;
  width: 24px;
  justify-content: center;
  border-radius: 4px;
}

.c20 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-top: 24px;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 0 0 auto;
  justify-content: space-between;
}

.c2 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  align-items: flex-start;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 0 0 auto;
  flex-wrap: wrap;
  column-gap: 12px;
  row-gap: 12px;
}

.c19 {
  flex: 0 0 auto;
  align-self: stretch;
  height: 12px;
}

.c6 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c6 g {
  fill: inherit;
  stroke: inherit;
}

.c6 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c6 *[stroke*='#'],
.c6 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c6 *[fill-rule],
.c6 *[FILL-RULE],
.c6 *[fill*='#'],
.c6 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c12 {
  margin-left: 12px;
  margin-right: 12px;
  margin-top: 6px;
  margin-bottom: 6px;
  font-size: 18px;
  line-height: 24px;
}

.c21 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  transition-property: color,background-color,border-color,box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;
}

.c21:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c21:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c21:focus >circle,
.c21:focus >ellipse,
.c21:focus >line,
.c21:focus >path,
.c21:focus >polygon,
.c21:focus >polyline,
.c21:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c21:focus ::-moz-focus-inner {
  border: 0;
}

.c21:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c21:focus:not(:focus-visible) >circle,
.c21:focus:not(:focus-visible) >ellipse,
.c21:focus:not(:focus-visible) >line,
.c21:focus:not(:focus-visible) >path,
.c21:focus:not(:focus-visible) >polygon,
.c21:focus:not(:focus-visible) >polyline,
.c21:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c21:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c8 {
  max-width: 100%;
}

.c14 {
  display: flex;
  flex-direction: row;
  align-items: center;
  user-select: none;
  -webkit-user-select: none;
  width: fit-content;
  cursor: pointer;
}

.c14:hover input:not([disabled])+div,
.c14:hover input:not([disabled])+span {
  border-color: #000000;
}

.c17 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

.c17:checked+span>span {
  left: calc( 48px - 24px );
  background: #7D4CDB;
}

.c16 {
  flex-shrink: 0;
}

.c7 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  width: 100%;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0, 0, 0, 0.33);
  border-radius: 4px;
  padding-left: 48px;
}

.c7:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c7:focus >circle,
.c7:focus >ellipse,
.c7:focus >line,
.c7:focus >path,
.c7:focus >polygon,
.c7:focus >polyline,
.c7:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c7:focus ::-moz-focus-inner {
  border: 0;
}

.c7::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c7::-moz-placeholder {
  color: #AAAAAA;
}

.c7:-ms-input-placeholder {
  color: #AAAAAA;
}

.c7 ::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c7::-moz-focus-inner {
  border: none;
  outline: none;
}

.c7:-moz-placeholder,
.c7::-moz-placeholder {
  opacity: 1;
}

.c4 {
  position: relative;
  width: 100%;
}

.c5 {
  position: absolute;
  display: flex;
  justify: center;
  top: 50%;
  transform: translateY(-50%);
  pointer-events: none;
  left: 12px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

@media only screen and (max-width: 768px) {

}

@media only screen and (max-width: 768px) {
  .c11 {
    margin-bottom: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c13 {
    border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  }
}

@media only screen and (max-width: 768px) {
  .c13 {
    padding: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c15 {
    margin-right: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c18 {
    border: solid 2px rgba(0, 0, 0, 0.15);
  }
}

@media only screen and (max-width: 768px) {
  .c20 {
    margin-top: 12px;
  }
}

@media only screen and (max-width: 768px) {
  .c2 {
    column-gap: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c19 {
    height: 6px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
    id="data"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <svg
              aria-label="Search"
              class="c6"
              viewBox="0 0 24 24"
            >
              <path
                d="m15 15 7 7-7-7zm-5.5 2a7.5 7.5 0 1 0 0-15 7.5 7.5 0 0 0 0 15z"
                fill="none"
                stroke="#000"
                stroke-width="2"
              />
            </svg>
          </div>
          <input
            aria-label="Search"
            autocomplete="off"
            class="c7"
            id="data--search"
            name="_search"
            type="search"
            value=""
          />
        </div>
      </div>
      <form
        class="c8"
      >
        <div
          class="c9"
        >
          <div
            class="c10"
          >
            <div
              class="c1"
            >
              <div
                class="c11 "
              >
                <label
                  class="c12"
                  for="data-name"
                >
                  name
                </label>
                <div
                  class="c13 "
                >
                  <div
                    class="c3 StyledCheckBoxGroup-sc-2nhc5d-0"
                    id="data-name"
                    role="group"
                  >
                    <label
                      class="c14"
                    >
                      <div
                        class="c15 c16"
                      >
                        <input
                          class="c17"
                          type="checkbox"
                        />
                        <div
                          class="c18 "
                        />
                      </div>
                      <span>
                        a
                      </span>
                    </label>
                    <div
                      class="c19"
                    />
                    <label
                      class="c14"
                    >
                      <div
                        class="c15 c16"
                      >
                        <input
                          class="c17"
                          type="checkbox"
                        />
                        <div
                          class="c18 "
                        />
                      </div>
                      <span>
                        b
                      </span>
                    </label>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <footer
            class="c20"
          >
            <button
              class="c21"
              type="submit"
            >
              Apply filters
            </button>
          </footer>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`DataFilters renders outside grommet wrapper 1`] = `
.c0 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 0 0 auto;
}

.c1 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  align-items: flex-start;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 0 0 auto;
  column-gap: 12px;
}

.c2 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c8 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 100%;
}

.c9 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  flex: 1 1;
  overflow: auto;
}

.c10 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-bottom: 12px;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
}

.c12 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  padding: 12px;
}

.c14 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-right: 12px;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  justify-content: center;
}

.c17 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  align-items: center;
  border: solid 2px rgba(0, 0, 0, 0.15);
  min-width: 0;
  min-height: 0;
  flex-direction: column;
  height: 24px;
  width: 24px;
  justify-content: center;
  border-radius: 4px;
}

.c19 {
  display: flex;
  box-sizing: border-box;
  max-width: 100%;
  margin-top: 24px;
  align-items: center;
  min-width: 0;
  min-height: 0;
  flex-direction: row;
  flex: 0 0 auto;
  justify-content: space-between;
}

.c18 {
  flex: 0 0 auto;
  align-self: stretch;
  height: 12px;
}

.c5 {
  display: inline-block;
  flex: 0 0 auto;
  width: 24px;
  height: 24px;
  fill: #666666;
  stroke: #666666;
}

.c5 g {
  fill: inherit;
  stroke: inherit;
}

.c5 *:not([stroke])[fill='none'] {
  stroke-width: 0;
}

.c5 *[stroke*='#'],
.c5 *[STROKE*='#'] {
  stroke: inherit;
  fill: none;
}

.c5 *[fill-rule],
.c5 *[FILL-RULE],
.c5 *[fill*='#'],
.c5 *[FILL*='#'] {
  fill: inherit;
  stroke: none;
}

.c11 {
  margin-left: 12px;
  margin-right: 12px;
  margin-top: 6px;
  margin-bottom: 6px;
  font-size: 18px;
  line-height: 24px;
}

.c20 {
  display: inline-block;
  box-sizing: border-box;
  cursor: pointer;
  font: inherit;
  text-decoration: none;
  margin: 0;
  background: transparent;
  overflow: visible;
  text-transform: none;
  border: 2px solid #7D4CDB;
  border-radius: 18px;
  color: #444444;
  padding: 4px 22px;
  font-size: 18px;
  line-height: 24px;
  background-color: #7D4CDB;
  color: #f8f8f8;
  border-radius: 18px;
  transition-property: color,background-color,border-color,box-shadow;
  transition-duration: 0.1s;
  transition-timing-function: ease-in-out;
}

.c20:hover {
  box-shadow: 0px 0px 0px 2px #7D4CDB;
}

.c20:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c20:focus >circle,
.c20:focus >ellipse,
.c20:focus >line,
.c20:focus >path,
.c20:focus >polygon,
.c20:focus >polyline,
.c20:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c20:focus ::-moz-focus-inner {
  border: 0;
}

.c20:focus:not(:focus-visible) {
  outline: none;
  box-shadow: none;
}

.c20:focus:not(:focus-visible) >circle,
.c20:focus:not(:focus-visible) >ellipse,
.c20:focus:not(:focus-visible) >line,
.c20:focus:not(:focus-visible) >path,
.c20:focus:not(:focus-visible) >polygon,
.c20:focus:not(:focus-visible) >polyline,
.c20:focus:not(:focus-visible) >rect {
  outline: none;
  box-shadow: none;
}

.c20:focus:not(:focus-visible) ::-moz-focus-inner {
  border: 0;
}

.c7 {
  max-width: 100%;
}

.c13 {
  display: flex;
  flex-direction: row;
  align-items: center;
  user-select: none;
  -webkit-user-select: none;
  width: fit-content;
  cursor: pointer;
}

.c13:hover input:not([disabled])+div,
.c13:hover input:not([disabled])+span {
  border-color: #000000;
}

.c16 {
  opacity: 0;
  -moz-appearance: none;
  width: 0;
  height: 0;
  margin: 0;
  cursor: pointer;
}

.c16:checked+span>span {
  left: calc( 48px - 24px );
  background: #7D4CDB;
}

.c15 {
  flex-shrink: 0;
}

.c6 {
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: none;
  -webkit-appearance: none;
  background: transparent;
  color: inherit;
  width: 100%;
  padding: 11px;
  font-weight: 600;
  margin: 0;
  border: 1px solid rgba(0, 0, 0, 0.33);
  border-radius: 4px;
  padding-left: 48px;
}

.c6:focus {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c6:focus >circle,
.c6:focus >ellipse,
.c6:focus >line,
.c6:focus >path,
.c6:focus >polygon,
.c6:focus >polyline,
.c6:focus >rect {
  outline: none;
  box-shadow: 0 0 2px 2px #6FFFB0;
}

.c6:focus ::-moz-focus-inner {
  border: 0;
}

.c6::-webkit-input-placeholder {
  color: #AAAAAA;
}

.c6::-moz-placeholder {
  color: #AAAAAA;
}

.c6:-ms-input-placeholder {
  color: #AAAAAA;
}

.c6 ::-webkit-search-decoration {
  -webkit-appearance: none;
}

.c6::-moz-focus-inner {
  border: none;
  outline: none;
}

.c6:-moz-placeholder,
.c6::-moz-placeholder {
  opacity: 1;
}

.c3 {
  position: relative;
  width: 100%;
}

.c4 {
  position: absolute;
  display: flex;
  justify: center;
  top: 50%;
  transform: translateY(-50%);
  pointer-events: none;
  left: 12px;
}

@media only screen and (max-width: 768px) {
  .c1 {
    column-gap: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c10 {
    margin-bottom: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c12 {
    border-bottom: solid 1px rgba(0, 0, 0, 0.33);
  }
}

@media only screen and (max-width: 768px) {
  .c12 {
    padding: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c14 {
    margin-right: 6px;
  }
}

@media only screen and (max-width: 768px) {
  .c17 {
    border: solid 2px rgba(0, 0, 0, 0.15);
  }
}

@media only screen and (max-width: 768px) {
  .c19 {
    margin-top: 12px;
  }
}

@media only screen and (max-width: 768px) {
  .c18 {
    height: 6px;
  }
}

<div
  class="c0"
  id="data"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <div
          class="c4"
        >
          <svg
            aria-label="Search"
            class="c5"
            viewBox="0 0 24 24"
          >
            <path
              d="m15 15 7 7-7-7zm-5.5 2a7.5 7.5 0 1 0 0-15 7.5 7.5 0 0 0 0 15z"
              fill="none"
              stroke="#000"
              stroke-width="2"
            />
          </svg>
        </div>
        <input
          aria-label="Search"
          autocomplete="off"
          class="c6"
          id="data--search"
          name="_search"
          type="search"
          value=""
        />
      </div>
    </div>
    <form
      class="c7"
    >
      <div
        class="c8"
      >
        <div
          class="c9"
        >
          <div
            class="c0"
          >
            <div
              class="c10 "
            >
              <label
                class="c11"
                for="data-name"
              >
                name
              </label>
              <div
                class="c12 "
              >
                <div
                  class="c2 StyledCheckBoxGroup-sc-2nhc5d-0"
                  id="data-name"
                  role="group"
                >
                  <label
                    class="c13"
                  >
                    <div
                      class="c14 c15"
                    >
                      <input
                        class="c16"
                        type="checkbox"
                      />
                      <div
                        class="c17 "
                      />
                    </div>
                    <span>
                      a
                    </span>
                  </label>
                  <div
                    class="c18"
                  />
                  <label
                    class="c13"
                  >
                    <div
                      class="c14 c15"
                    >
                      <input
                        class="c16"
                        type="checkbox"
                      />
                      <div
                        class="c17 "
                      />
                    </div>
                    <span>
                      b
                    </span>
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
        <footer
          class="c19"
        >
          <button
            class="c20"
            type="submit"
          >
            Apply filters
          </button>
        </footer>
      </div>
    </form>
  </div>
</div>
`;
